### PR TITLE
Add layui w/ git auto-update

### LINKED ISSUE #11369: [Request] Add layui
**Library name:** layui
**Git repository url:** https://github.com/sentsin/layui/
**License:** MIT
**Official homepage:** http://www.layui.com/

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/45892398-request-add-layui?utm_campaign=plugin&utm_content=tracker%2F32893&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F32893&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Billy4195**
close #11369, cc @sentsin.

Pull request for issue: #11369
Related issue(s): 

Checklist for **Pull request** or **lib adding request issue** follows the conventions.

Note that if you are using a distribution purpose repository/package, please also provide the url and other related info like popularity of the source code repo/package.

# Profile of the lib
 * Git repository (required): https://github.com/sentsin/layui
 * Official website (optional, not the repository):
 * NPM package url (optional): [layui-src](https://www.npmjs.com/package/layui-src)(v2.1.5 ~ v2.2.45) [layui-layer](https://www.npmjs.com/package/layui-layer)(v1.0.0 ~ v1.0.9)
 * License and its reference: MIT
 * GitHub / Bitbucket popularity (required):
   - Count of watchers: 707
   - Count of stars: 9807
   - Count of forks: 2823
 * NPM download stats (optional):
   - Downloads in the last day: 37
   - Downloads in the last week: 195
   - Downloads in the last month: 925

# Essential checklist
 * [ ] I'm the author of this library
   * [ ] I would like to add link to the page of this library on CDNJS on website / readme
 * [x] This lib was not found on cdnjs repo
 * [x] No already exist / duplicated issue and PR
 * [x] The lib has notable popularity
   * [x] More than 200 [Stars / Watchers / Forks] on [GitHub / Bitbucket]
   * [ ] More than 800 downloads stats per month on npm registry
 * [x] Project has public repository on famous online hosting platform (or been hosted on npm)

# Auto-update checklist
 * [x] Has valid tags for each versions (for git auto-update)
 * [x] Auto-update setup
 * [x] Auto-update target/source is valid.
 * [x] Auto-update filemap is correct.

# Git commit checklist
 * [x] The first line of commit message is less then 50 chars, be clean and clear, easy to understand.
 * [x] The parent of the commit(s) in the PR is not old than 3 days.
 * [x] Pull request is sending from a non-master branch with meaningful name.
 * [x] Separate unrelated changes into different commits.
 * [x] Use rebase to squash/fixup dummy/unnecessary commits into only one commit.
 * [x] Close corresponding issue in commit message
 * [x] Mention related issue(s), people in commit message, comment.

### CODE REVIEW

**Comment by Billy4195** [comment]
@sashberd Done.

File: ajax/libs/layui/package.json [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  "keywords": [
+    "layui",
+    "ui"
+  ],
+  "autoupdate": {
+    "source": "git",
+    "target": "https://github.com/sentsin/layui.git",
+    "fileMap": [
+      {
+        "basePath": "dist",
```
Comment (sashberd) [inline ajax/libs/layui/package.json]: In versions below 2.0.0 there are no `dist `folder.There are a `build `folder, thus please another pattern here to fetch also `build `directory

**Comment by sashberd** [comment]
Also I think this is the [npm ](https://www.npmjs.com/package/layui-layer)of current package. If you agree please add info to the PRs info from there.
Also please add version [2.2.2 ](https://libraries.io/npm/layui-src/2.2.2 )and [2.1.61](https://libraries.io/npm/layui-src/2.1.61) from npm manually due to non existence on git repo. 
Thanks

**Comment by sashberd** [comment]
Another things that layui has a lot of plugins e.g. [layui-layer](https://www.npmjs.com/package/layui-layer) may be we could add requests to cdnjs in order to fetch full library scope. FY discretion

**Comment by Billy4195** [comment]
@sashberd I have add npm info in PR list.
I can't find v2.2.2 & v2.1.61 on npm page, but I add old version files manually.
And I can't understand what the plugins means, I think that has nothing to do with this PR, this PR already contain the essential files to build up, maybe we can open another issue to discuss those plugins.

**Comment by sashberd** [comment]
@Billy4195 As I understand you added old version manually not via configuration... this is OK, but via configuration this is cleaner wasy and easy to review... Lets wait another reviewers opinions for such adds.

As for the version on npm I sent you direct links in previous message. Moreover, I suggest you to use [](https://libraries.io) page to use and find libraries on npm

> Also please add version 2.2.2 and 2.1.61 from npm manually due to non existence on git repo.

**Comment by sentsin** [comment]
@sashberd @Billy4195 
I think layui v2.2.2 and v2.1.61 can be ignored, I forgot to publish to npm, and very few people use these two versions. thx

File: ajax/libs/layui/package.json [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  "name": "layui",
+  "filename": "layui.js",
+  "version": "2.2.45",
+  "description": "经典模块化前端框架",
+  "license": "MIT",
+  "repository": {
+    "type": "git",
+    "url": "git+ssh://git@github.com/sentsin/layui.git"
+  },
+  "author": "贤心",
```
Comment (extend1994) [inline ajax/libs/layui/package.json]: Here should be in English too, you can use info in bower.json from upstream https://github.com/sentsin/layui/blob/master/bower.json#L6.
When you write package.json for cdnjs library. package.json and bower.json from upstream are both good reference :)

File: ajax/libs/layui/package.json [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  "homepage": "http://www.layui.com",
```
Comment (extend1994) [inline ajax/libs/layui/package.json]: Please check if https workable when url was used. Please use https here.

File: ajax/libs/layui/package.json [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  "keywords": [
+    "layui",
+    "ui"
```
Comment (extend1994) [inline ajax/libs/layui/package.json]: (Optional) More keywords may help users find this library.

**Comment by extend1994** [comment]
@sentsin Thanks for attention here. 
I suppose https://www.npmjs.com/package/layui-layer is not from you or your team because I didn't see the guy on the contributors page on GitHub. Could you please help us check this? And I think there is no plugin of lay?

**Comment by sentsin** [comment]
@extend1994 Right. That one is not ours. This one is ours: https://www.npmjs.com/package/layui-src

File: ajax/libs/layui/package.json [hunk 5]
```diff
@@ -0,0 +1,35 @@
+{
+  "name": "layui",
+  "filename": "layui.js",
+  "version": "2.2.45",
+  "description": "经典模块化前端框架",
```
[the remark this answers is not in the document]
Comment (sentsin) [reply ajax/libs/layui/package.json]: Done

File: ajax/libs/layui/package.json (hunk 2 above)
Comment (sentsin) [reply ajax/libs/layui/package.json]: Done

File: ajax/libs/layui/package.json (hunk 3 above)
Comment (sentsin) [reply ajax/libs/layui/package.json]: Done

File: ajax/libs/layui/package.json (hunk 4 above)
Comment (sentsin) [reply ajax/libs/layui/package.json]: okay thanks for suggest

**Comment by Billy4195** [comment]
@extend1994 I finished the change. 
@sentsin thanks for helping us :smile:

**Comment by sentsin** [comment]
@Billy4195 哈哈，我應該謝謝妳們才對，終於把 layui 收錄了。好吧，我還是沒忍住說中文...

**Comment by extend1994** [comment]
@sentsin It's fine you typed in Chinese. For some reason, we will still type in English though.
cdnjs hopes to provide versions as complete as possible so we would add them basically. Appreciate your opinion on it 😄, reviewers may discuss it later!

**Comment by Billy4195** [comment]
@sufuf3 Got it. Thanks for modifying the commit message.

**Comment by PeterDaveHello** [comment]
@sentsin do you have suggestions for my review above ⬆️,  please?  Thanks a lot.

**Comment by sentsin** [comment]
@PeterDaveHello 抱歉，現在才看到。
在 1.x 的版本中，的確是存在 build 目錄的，2.x 開始改為 dist 替代了。但我覺得問題不大，其實我們可以忽略 1.x 的版本，現在已經很少人在使用 1.x 了。妳們覺得呢？ 😄

**Comment by PeterDaveHello** [comment]
@sentsin So technically build should be the same as dist? I think they didn't look so same. We can use just dist for auto-update config and I'd manually add v1.x, adding the old versions can help users migrate to use CDN easier.

BTW, no offense but I believe that we should use English here.

**Comment by sentsin** [comment]
@PeterDaveHello 👍  Yes, actually they are the same apart from some tiny differences. I am wondering if we can abandon the version of 1.x. If the workload is too huge, then I think there is no need to add v1.x manually.

Is it OK to add layui 2.x only? 😅

**Comment by PeterDaveHello** [comment]
@sentsin it's fine to add v2.x only, though IMO I don't see the dificually to add v1.x yet, so I'd prefer to add them all.

Would you please tell us about the difference about "dist" and "build" in v2.x?

**Comment by sentsin** [comment]
@PeterDaveHello OK. The main difference is that layui.all.js is moved from the directory named / build / lay / dest / to the root directory named / dist /. Thus, it is at the same level as the layui.js directory. Besides, the / lay / dest / directory is also removed. The rest of them are still the same. 
Reference: https://github.com/sentsin/layui/tree/v1/build

**Comment by sashberd** [comment]
@PeterDaveHello As I understood from conversation, we need to add v2.x via auto-update config and v1.x manually?

**Comment by sashberd** [comment]
Moved to #13721

**Pull request closed without merging**